Fix strides format for data format in contrib.layers.separable convolution2d

jcastagneri opened the pull request:
Addresses feature request from issue #10432. Continuation of feature implementation--initial changes did not address the case where num_outputs = None, in which case strides and the input channels must be reformulated according to data format. The previous commit is  rmlarsen@d52e15a.

I created a previous PR #12120 for a similar issue that @rmlarsen fixed in the above commit, but the strides were still mismatched to data format--I thought it would be cleaner if I just made a new PR.

Edit--I thought it would be cleaner because my CLA would be recognized, but it still hasn't. See the comment below.

jcastagneri committed:
Strides must change form if data format is switched

jcastagneri committed:
Fixed stride mismatch on channels first data format for separable conv

jcastagneri committed:
Fixed sepconv2d test and added stride testing

tensorflow-jenkins commented:
Can one of the admins verify this patch?

jcastagneri commented:
I don't know why my CLA is failing... The commits list my CLA email: joe@ctxrobotics.com. When I click the redirect link from the CLA bot, it shows I am covered under a corporate CLA. As far as I am aware, I have taken the necessary steps to be certified under the CLA. Anybody know what I am missing here?

jcastagneri committed:
Merge branch 'master' into fix-strides-format-for-data-format-in-contrib.layers.separable_convolution2d

jcastagneri commented:
@rmlarsen Is there any movement on this issue? It's really just a one line fix plus the test update, but until it is merged the channels first separable convolution from contrib is wrong.

jhseu commented:
Keep in mind that we can't look at the code until the CLA is resolved. It looks like you have commits without e-mails in the list of commits. Perhaps resolve that?

drpngx commented:
@jcastagneri any luck on the CLA? We'll have to close this PR otherwise. Thanks.

sb2nov commented:
@jcastagneri It might be useful to try checking the CLA incognito to make sure it is validating the correct account.

drpngx commented:
Ah, @jcastagneri you have signed via a company. You need to have the company on your public github profile for the CLA checking to work. Thanks.

jcastagneri commented:
Thanks for the fix @drpngx. I added the company name, but the company is not actually on Github. Will this still block the CLA bot? I just added 'ctxrobotics' in the company section of my bio, as this is what I believe to be the company name registered to the CLA.

martinwicke commented:
@willnorris for knowledge of how exactly company CLAs work.

I believe your company has to have the email on file with the CLA, and that email must be the same as the one used in the commits.

willnorris commented:
@jcastagneri: it looks like you need to add your ctx robotics email [to your github account](https://help.github.com/articles/adding-an-email-address-to-your-github-account/).

jcastagneri commented:
I have just added the email. Do I have to kickoff the CLA bot manually or will it check automatically?

drpngx commented:
Great, I think it's ready for review now.

@rmlarsen it looks like this is something you might have some prior knowledge about?

sguada commented on tensorflow/contrib/layers/python/layers/layers.py:
```diff
@@ -2563,7 +2563,7 @@ def separable_convolution2d(
           trainable=trainable,
           collections=weights_collections)
-      strides = [1, stride_h, stride_w, 1]
+      strides = [1, 1, stride_h, stride_w] if data_format.startswith('NC') else [1, stride_h, stride_w, 1]
```
Can you add a test where this change is tested? i.e. when num_outputs = None but depth_multiplier is Not.

jcastagneri replied:
@sguada As far as I am aware, my test already does this check (`testConvNCHW` in `layers_test.py`), which is a part of this PR. Specifically, the test will assert correct output shape size for NCHW data input twice: Once with `num_outputs = None`, and with `num_outputs=5`, a non-None test value. Both tests are done with `depth_multiplier = 2`, which I believe satisfies the needed tests for this change. Let me know if I have missed something!

sguada reviewed (approved):
Yeah I overlooked that.

martinwicke commented:
Jenkins. test this please.
